File: get_weather_data.py
import json
import psycopg2
from psycopg2 import sql
from dotenv import load_dotenv
import os

# Load environment variables from .env file
load_dotenv()

# Database connection details from environment variables
host = os.getenv("DB_HOST")
port = os.getenv("DB_PORT")
dbname = os.getenv("DB_NAME")
user = os.getenv("DB_USER")
password = os.getenv("DB_PASSWORD")

try:
    # Ensure environment variables are loaded
    if not all([host, port, dbname, user, password]):
        raise ValueError("Some environment variables for DB connection are missing.")

    # Connect to the PostgreSQL database
    conn = psycopg2.connect(
        dbname=dbname,
        user=user,
        password=password,
        host=host,
        port=port
    )

    # Use a context manager for the cursor
    with conn.cursor() as cursor:
        # Set the region_id to retrieve
        region_id = 2

        # SQL query to retrieve the latest weather data for the specified region_id
        query = sql.SQL(""" 
            SELECT temperature, wind_speed, humidity, current_instruction 
            FROM weather_data 
            WHERE region_id = %s 
            AND id = (SELECT MAX(id) FROM weather_data WHERE region_id = %s); 
        """)

        # Execute the query
        cursor.execute(query, (region_id, region_id))

        # Fetch the result
        data = cursor.fetchone()

        # Check if data was found and output in JSON format
        if data:
            ordered_data = {
                'temperature': data[0],
                'wind_speed': data[1],
                'humidity': data[2],
                'current_instruction': data[3],
            }
            print(json.dumps(ordered_data))
        else:
            print(json.dumps({"error": "No data found"}))

except Exception as e:
    # Handle connection and query errors
    print(json.dumps({"error": f"Connection or query failed: {str(e)}"}))

finally:
    # Close the connection
    if conn:
        conn.close()

File: get_weather_data.php
<?php
// Enable error reporting for troubleshooting
ini_set('display_errors', 1);
ini_set('display_startup_errors', 1);
error_reporting(E_ALL);

// Database connection details for PostgreSQL
$host = "dpg-csdvkr3v2p9s73b2bkog-a";
$port = "5432";
$dbname = "preventers";
$user = "preventers_user";
$password = "xR7u0DpZsQRdyArWZGmrCQ7zMrH53lQp";

try {
    // Set up the DSN for PostgreSQL
    $dsn = "pgsql:host=$host;port=$port;dbname=$dbname";
    $pdo = new PDO($dsn, $user, $password);

    // Set error mode to exception for easier debugging
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);

    // Set the region_id to retrieve
    $region_id = 2;

    // SQL query to retrieve the latest weather data for the specified region_id
    $sql = "
    SELECT temperature, wind_speed, humidity, current_instruction
    FROM weather_data 
    WHERE region_id = :region_id 
    AND id = (SELECT MAX(id) FROM weather_data WHERE region_id = :region_id);
    ";

    // Prepare and execute the statement
    $stmt = $pdo->prepare($sql);
    $stmt->bindParam(':region_id', $region_id, PDO::PARAM_INT);
    $stmt->execute();
    $data = $stmt->fetch(PDO::FETCH_ASSOC);

    // Check if data was found
    if ($data) {
        // Output data in JSON format
        $ordered_data = [
            'temperature' => $data['temperature'],
            'wind_speed' => $data['wind_speed'],
            'humidity' => $data['humidity'],
            'current_instruction' => $data['current_instruction'],
        ];
        echo json_encode($ordered_data);
    } else {
        echo json_encode(["error" => "No data found"]);
    }
} catch (PDOException $e) {
    // Handle connection and query errors
    echo json_encode(["error" => "Connection or query failed: " . $e->getMessage()]);
}
?>
